qutebrowser/completion/strategies/strategy.py

```diff
@@ -0,0 +1,48 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+# Copyright 2014-2020 Florian Bruhin (The Compiler) <mail@qutebrowser.org>
+#
+# This file is part of qutebrowser.
+#
+# qutebrowser is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# qutebrowser is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with qutebrowser.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Classes that return miscellaneous completion models."""
+
+from typing import Sequence, Optional, Any, Callable, Tuple
+from abc import ABC, abstractmethod
+
+from qutebrowser.completion.completionmodel import CompletionModel
+from qutebrowser.completion.completer import CompletionInfo
+
+
+class DeletionUnsupportedError(Exception):
+    """Raises when the completion strategy does not support deletion."""
+
+
+class CompletionStrategy(ABC):
+    COLUMN_WIDTHS: Tuple[int, int, int] = (20, 70, 10)
+
+    def __init__(self):
+        self.model = CompletionModel(column_widths=self.COLUMN_WIDTHS)
```

Comment:
`model` doesn't need to be a field. Let's build a new one from scratch in `populate`.

Reply:
I wonder what @The-Compiler's thoughts on this are? Since this design was proposed by him. We can do as you suggested unless @The-Compiler has some use in mind for this API design. After all, the entire reason we went for this refactor was to extend the functionality of these classes.

Reply:
Hmm, yes this does match the initial suggestion. It isn't a big deal either way, but I'm not sure why `model` needs to be a member. I'd prefer a less stateful option like: 

1.
```
class CompletionStrategy(ABC):
    @abstractmethod
    def populate(model):
        # implemented by subclass

    def __call__(...):
        model = CompletionModel()
        _populate(model)
        return model
```

2.
```
class CompletionStrategy(ABC):
    @abstractmethod
    def populate() -> model:
        # implemented by subclass, which should construct a new model in the method
```

I might be misunderstanding, but I think that is `model` is a member then `populate()` needs to be responsible for clearing it somehow.

Reply:
Just so that we are in the same page here, is this how you envision the API to be:
```
class AbstractCompletionStrategy:
    COLUMN_WIDTHS: Tuple[int, int, int] = (20, 70, 10)

    def complete(self, *args: Any, info: CompletionInfo) -> CompletionModel:
        model = CompletionModel(column_widths=self.COLUMN_WIDTHS)
        return self._populate(*args, info, model)

    def _populate(self, *args: str, info: CompletionInfo,
                  model: CompletionModel) -> CompletionModel:
        raise NotImplementedError

    @classmethod
    def delete(cls, data: Sequence[str]) -> None:
        raise DeletionUnsupportedError(f"{cls} does not support deletion")
```

One thing to note here is that I will need to pass the info as well as the model to populate since populate uses info a lot of times. I'm guessing args will also be necessary? Let me know if this makes sense

Reply:
Hey @The-Compiler, just a friendly reminder in case you missed this. Let me know if you agree with the above design and I'll go ahead and implement it. No worries if you are busy though.

Reply:
LGTM, though I'll let @The-Compiler make the final call.

Reply:
I missed this indeed, apologies! I'm afraid I won't have time for a deeper look at this in the next 2 weeks or so. Seems fine from a (very) quick look, though.

Reply:
No worries! Thanks for taking the time. When you get some time, can you please take a look at this as well: https://github.com/qutebrowser/qutebrowser/pull/6217#discussion_r602877953.